crypto/crypto.go

```diff
@@ -0,0 +1,163 @@
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package crypto
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/crypto/storage"
+)
+
+// Config holds the values for the crypto engine
+type Config struct {
+	Storage string
+	Fspath  string
+}
+
+func (cc Config) getFSPath() string {
+	if cc.Fspath == "" {
+		return DefaultCryptoConfig().Fspath
+	}
+
+	return cc.Fspath
+}
+
+// DefaultCryptoConfig returns a Config with sane defaults
+func DefaultCryptoConfig() Config {
+	return Config{
+		Storage: "fs",
+		Fspath:  "./",
+	}
+}
+
+// Crypto holds references to storage and needed config
+type Crypto struct {
+	Storage    storage.Storage
+	Config     Config
+	configOnce sync.Once
+	configDone bool
+}
+
+type opaquePrivateKey struct {
+	publicKey crypto.PublicKey
+	signFn    func(io.Reader, []byte, crypto.SignerOpts) ([]byte, error)
+}
+
+// Public returns the public key
+func (k opaquePrivateKey) Public() crypto.PublicKey {
+	return k.publicKey
+}
+
+// Sign signs some data with the signer
+func (k opaquePrivateKey) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	return k.signFn(rand, digest, opts)
+}
+
+// Shutdown stops the certificate monitors
+func (client *Crypto) Shutdown() error {
+	return nil
+}
+
+// GetPrivateKey returns the specified private key. It can be used for signing, but cannot be exported.
+func (client *Crypto) GetPrivateKey(kid string) (crypto.Signer, error) {
+	priv, err := client.Storage.GetPrivateKey(kid)
+	if err != nil {
+		return nil, err
+	}
+	return opaquePrivateKey{publicKey: priv.Public(), signFn: priv.Sign}, nil
+}
+
+var instance *Crypto
+
+var oneBackend sync.Once
+
+// Instance returns the same instance of Crypto every time
+func Instance() *Crypto {
+	if instance != nil {
+		return instance
+	}
+	oneBackend.Do(func() {
+		instance = &Crypto{
+			Config: DefaultCryptoConfig(),
+		}
+	})
+	return instance
+}
+
+// Configure loads the given configurations in the engine. Any wrong combination will return an error
+func (client *Crypto) Configure() error {
+	var err error
+	client.configOnce.Do(func() {
+		if core.NutsConfig().Mode() != core.ServerEngineMode {
+			return
+		}
+		if err = client.doConfigure(); err == nil {
+			client.configDone = true
+		}
+	})
+	return err
+}
+
+func (client *Crypto) doConfigure() error {
+	if client.Config.Storage != "fs" && client.Config.Storage != "" {
+		return errors.New("only fs backend available for now")
+	}
+	var err error
+	if client.Storage, err = storage.NewFileSystemBackend(client.Config.getFSPath()); err != nil {
+		return err
+	}
+	return nil
+}
+
+// New generates a new key pair. If a key pair with the same identifier already exists, it is overwritten.
```

Comment:
"If a key pair with the same identifier already exists, it is overwritten."

Why not error? I think it should never be allowed to happen?

Reply:
wrong docs. Changed it to: should be handled by storage and bad practise to re-use a kid.